File: SortingMethods/quickSort.py
def quickSortArr(array):
    quickSort(array, 0, len(array) - 1)

def quickSort(array, low, high):
    if low < high:
        splitIndex = partition(array, low, high)

        quickSort(array, low, splitIndex - 1)
        quickSort(array, splitIndex + 1, high)

def partition(array, low, high):
    pivot = array[high]
    i = (low - 1)

    for j in range(low, high):
        if array[j] < pivot:
            i += 1
            array[i], array[j] = array[j], array[i]

    array[i+1], array[high] = array[high], array[i+1]
    return i + 1

File: SortingMethods/src/QuickSort.java
public class QuickSort<T extends Comparable<T>> implements SortingAlgorithm<T>{

    public void sort(T arr[]){
        quicksort(arr,0,arr.length-1);
    }

    void swap (T arr[],int i, int j){
        T temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    int partition(T arr[], int low, int high)
    {
        T pivot = arr[high];
        int i = (low-1);
        for (int j=low; j<high; j++)
        {
            if (arr[j].compareTo(pivot)<0)
            {
                i++;
                swap(arr,i,j);
            }
        }

        swap(arr,i+1,high);

        return i+1;
    }

    void quicksort(T arr[], int low, int high)
    {
        if (low < high)
        {
            int pi = partition(arr, low, high);

            quicksort(arr, low, pi-1);
            quicksort(arr, pi+1, high);
        }
    }


}
